apps/gdalalg_vector_geom_make_valid.cpp

```diff
@@ -0,0 +1,144 @@
+/******************************************************************************
+ *
+ * Project:  GDAL
+ * Purpose:  "gdal vector geom make-valid"
+ * Author:   Even Rouault <even dot rouault at spatialys.com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2025, Even Rouault <even dot rouault at spatialys.com>
+ *
+ * SPDX-License-Identifier: MIT
+ ****************************************************************************/
+
+#include "gdalalg_vector_geom_make_valid.h"
+
+#include "gdal_priv.h"
+#include "ogrsf_frmts.h"
+
+//! @cond Doxygen_Suppress
+
+#ifndef _
+#define _(x) (x)
+#endif
+
+/************************************************************************/
+/*                 GDALVectorGeomMakeValidAlgorithm()                   */
+/************************************************************************/
+
+GDALVectorGeomMakeValidAlgorithm::GDALVectorGeomMakeValidAlgorithm(
+    bool standaloneStep)
+    : GDALVectorGeomAbstractAlgorithm(NAME, DESCRIPTION, HELP_URL,
+                                      standaloneStep, m_opts)
+{
+    AddArg("keep-lower-dim", 0,
+           _("Keep components of lower dimension after MakeValid()"),
+           &m_opts.m_keepLowerDim);
+}
+
+#ifdef HAVE_GEOS
+
+namespace
+{
+
+/************************************************************************/
+/*                  GDALVectorGeomMakeValidAlgorithmLayer               */
+/************************************************************************/
+
+class GDALVectorGeomMakeValidAlgorithmLayer final
+    : public GDALVectorGeomOneToOneAlgorithmLayer<
+          GDALVectorGeomMakeValidAlgorithm>
+{
+  protected:
+    using GDALVectorGeomOneToOneAlgorithmLayer::TranslateFeature;
+
+    std::unique_ptr<OGRFeature>
+    TranslateFeature(std::unique_ptr<OGRFeature> poSrcFeature) const override;
+
+  public:
+    GDALVectorGeomMakeValidAlgorithmLayer(
+        OGRLayer &oSrcLayer,
+        const GDALVectorGeomMakeValidAlgorithm::Options &opts)
+        : GDALVectorGeomOneToOneAlgorithmLayer<
+              GDALVectorGeomMakeValidAlgorithm>(oSrcLayer, opts)
+    {
+    }
+};
+
+/************************************************************************/
+/*                          TranslateFeature()                          */
+/************************************************************************/
+
+std::unique_ptr<OGRFeature>
+GDALVectorGeomMakeValidAlgorithmLayer::TranslateFeature(
+    std::unique_ptr<OGRFeature> poSrcFeature) const
+{
+    const int nGeomFieldCount = poSrcFeature->GetGeomFieldCount();
+    for (int i = 0; i < nGeomFieldCount; ++i)
+    {
+        if (IsSelectedGeomField(i))
+        {
+            auto poGeom =
+                std::unique_ptr<OGRGeometry>(poSrcFeature->StealGeometry(i));
+            if (poGeom)
```

Comment:
Pass through features that are (a) non-polygonal or (b) already valid ?

Reply:
fixed